### PR TITLE
PLT-3484 OAuth2 Service Provider

### PR DESCRIPTION
**Pull request by enahum**
#### Summary
This enables Mattermost to act as a OAuth2 Service Provider.

This PR includes:
* System Console settings for OAuth2 Service Provider
* Ability to add and delete OAuth2 apps from backstage
* Authenticate using OAuth2
* Issue a Token with a RefreshToken
* Sessions created by OAuth2 have the name of the App used to create the session

A [sample app](https://github.com/enahum/mattermost-sample-app) can be used to test the functionality.

#### Ticket Link
https://mattermost.atlassian.net/browse/PLT-3484

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [x] Added API documentation (required for all new APIs)
- [x] Has driver changes that have been merged and package.json updated
- [ ] Has enterprise changes (please link)
- [x] Has UI changes
- [x] Includes text changes and localization files updated
- [x] Touches critical sections of the codebase (auth, upgrade, etc.)
- [x] Has database changes

### CODE REVIEW

**Comment by jasonblais** [comment]
Update: Initial tests done on Windows 10, Chrome:
 - UI text updates have passed PM review, and will be sent to dev along with other proposed updates
 - will ask Asaad to take a look at potential UI tweaks + updating a field in the account settings modal

**Comment by jasonblais** [comment]
+1 from PM: 

Including previous tests, have tested on Windows 10, Chrome:
 - System Console settings
 - Testing UI in each screen in **Main Menu > Integrations**
 - Adding invalid parameters for OAuth 2.0 (to test error messages)
 - Establishing a successful connection with the application added to Active Sessions

A follow-up ticket [here](https://mattermost.atlassian.net/browse/PLT-3745) to add an option for an end user to deactivate an OAuth 2.0 app from Account Settings.

**Comment by hmhealey** [comment]
+1 Should we do a group review for this since it's pretty big? I looked over the serverside code the best I could, but there was a fair bit of stuff I couldn't quite follow

**Comment by enahum** [comment]
@hmhealey we can do that no problem

**Comment by crspeller** [comment]
Adding the work in progress label because we are going to await the group review.

**Comment by crspeller** [comment]
+1 Big and complicated though. 
Going to wait for @coreyhulen to give approval.

File: store/sql_oauth_store.go [hunk 1]
```diff
@@ -34,21 +36,36 @@ func NewSqlOAuthStore(sqlStore *SqlStore) OAuthStore {
...
 func (as SqlOAuthStore) UpgradeSchemaIfNeeded() {
+	as.CreateColumnIfNotExists("OAuthApps", "IsTrusted", "tinyint(1)", "boolean", "0")
+	as.CreateColumnIfNotExists("OAuthApps", "IconURL", "varchar(512)", "varchar(512)", "")
+	as.CreateColumnIfNotExists("OAuthAccessData", "ClientId", "varchar(26)", "varchar(26)", "")
+	as.CreateColumnIfNotExists("OAuthAccessData", "UserId", "varchar(26)", "varchar(26)", "")
+	as.CreateColumnIfNotExists("OAuthAccessData", "ExpiresAt", "bigint", "bigint", "0")
+
+	// ADDED for 3.3 REMOVE for 3.7
+	if as.DoesColumnExist("OAuthAccessData", "AuthCode") {
+		as.GetMaster().Exec("ALTER TABLE OAuthAccessData DROP INDEX idx_oauthaccessdata_auth_code")
```
Comment (coreyhulen) [inline store/sql_oauth_store.go]: Does this work for both mysql and postgress.  I wonder is we should use `RemoveIndexIfExists()` in sql_store.go

File: store/sql_oauth_store.go [hunk 2]
```diff
@@ -34,21 +36,36 @@ func NewSqlOAuthStore(sqlStore *SqlStore) OAuthStore {
...
+		as.GetMaster().Exec("ALTER TABLE OAuthAccessData DROP COLUMN AuthCode")
```
Comment (coreyhulen) [inline store/sql_oauth_store.go]: maybe used `RemoveColumnIfExists()`

**Comment by enahum** [comment]
@coreyhulen suggestions added

**Comment by coreyhulen** [comment]
+1

**Pull request merged**